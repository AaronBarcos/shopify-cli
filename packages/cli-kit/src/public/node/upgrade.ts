import {currentProcessIsGlobal, inferPackageManagerForGlobalCLI} from './is-global.js'
import {packageManagerFromUserAgent} from './node-package-manager.js'
import {outputContent, outputToken} from './output.js'

/**
 * Utility function for generating an install command for the user to run
 * to install an updated version of Shopify CLI.
 * @returns A string with the command to run.
 */
export function cliInstallCommand(): string {
  const isGlobal = currentProcessIsGlobal()
  let packageManager = packageManagerFromUserAgent() ?? inferPackageManagerForGlobalCLI()
  if (packageManager === 'unknown') packageManager = 'npm'

  if (packageManager === 'yarn') {
    return `yarn ${isGlobal ? 'global ' : ''}add @shopify/cli@latest`
  } else {
    return `${packageManager} i ${isGlobal ? '-g ' : ''}@shopify/cli@latest`
  }
}

/**
 * Generates a message to remind the user to update the CLI.
 *
 * @param version - The version to update to.
 * @returns The message to remind the user to update the CLI.
 */
export function getOutputUpdateCLIReminder(version: string): string {
  return outputContent`ðŸ’¡ Version ${version} available! Run ${outputToken.genericShellCommand(cliInstallCommand())}`
    .value
}
