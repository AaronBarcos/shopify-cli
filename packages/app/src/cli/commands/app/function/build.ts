import {inFunctionContext, functionFlags} from '../../../services/function/common.js'
import {buildFunctionExtension} from '../../../services/build/extension.js'
import {appFlags} from '../../../flags.js'
import AppCommand from '../../../utilities/app-command.js'
import {globalFlags} from '@shopify/cli-kit/node/cli'
import {renderSuccess} from '@shopify/cli-kit/node/ui'

export default class FunctionBuild extends AppCommand {
  static summary = 'Compile a function to wasm.'

  static descriptionWithMarkdown = `Compiles the function in your current directory to WebAssembly (Wasm) for testing purposes.`

  static description = this.descriptionWithoutMarkdown()

  static flags = {
    ...globalFlags,
    ...appFlags,
    ...functionFlags,
  }

  public async run() {
    const {flags} = await this.parse(FunctionBuild)
    await inFunctionContext({
      path: flags.path,
      userProvidedConfigName: flags.config,
      callback: async (app, ourFunction) => {
        await buildFunctionExtension(ourFunction, {
          app,
          stdout: process.stdout,
          stderr: process.stderr,
          useTasks: true,
          environment: 'production',
        })
        renderSuccess({headline: 'Function built successfully.'})
      },
    })
  }
}
