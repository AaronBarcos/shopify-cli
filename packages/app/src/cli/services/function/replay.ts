import {ensureConnectedAppFunctionContext} from '../generate-schema.js'
import {AppInterface} from '../../models/app/app.js'
import {ExtensionInstance} from '../../models/extensions/extension-instance.js'
import {FunctionConfigType} from '../../models/extensions/specifications/function.js'
import {selectFunctionRunPrompt} from '../../prompts/function/replay.js'

import {setupExtensionWatcher} from '../dev/extension/bundler.js'
import {joinPath} from '@shopify/cli-kit/node/path'
import {readFile} from '@shopify/cli-kit/node/fs'
import {getLogsDir} from '@shopify/cli-kit/node/logs'
import {exec} from '@shopify/cli-kit/node/system'
import {AbortError, FatalError} from '@shopify/cli-kit/node/error'
import {AbortController} from '@shopify/cli-kit/node/abort'

import {outputWarn} from '@shopify/cli-kit/node/output'
import {renderFatalError} from '@shopify/cli-kit/node/ui'
import {readdirSync} from 'fs'

const LOG_SELECTOR_LIMIT = 100

interface ReplayOptions {
  app: AppInterface
  extension: ExtensionInstance<FunctionConfigType>
  apiKey?: string
  stdout?: boolean
  path: string
  json: boolean
  watch: boolean
}

export interface FunctionRunData {
  shop_id: number
  api_client_id: number
  payload: {
    input: unknown
    input_bytes: number
    output: unknown
    output_bytes: number
    function_id: string
    export: string
    logs: string
    fuel_consumed: number
  }
  log_type: string
  cursor: string
  status: string
  source: string
  source_namespace: string
  log_timestamp: string
  identifier: string
}

export async function replay(options: ReplayOptions) {
  const {watch, extension, app} = options
  const abortController = new AbortController()

  try {
    const {apiKey} = await ensureConnectedAppFunctionContext(options)
    const functionRunsDir = joinPath(getLogsDir(), apiKey)
    const functionRuns = await getFunctionRunData(functionRunsDir, options.extension.handle)

    const selectedRun = await selectFunctionRunPrompt(functionRuns)

    if (selectedRun === undefined) {
      throw new AbortError(`No logs found in ${functionRunsDir}`)
    }

    const {input, export: runExport} = selectedRun.payload
    await runFunctionRunnerWithLogInput(extension, options, JSON.stringify(input), runExport)

    if (watch) {
      await setupExtensionWatcher({
        extension,
        app,
        stdout: process.stdout,
        stderr: process.stderr,
        onChange: async () => {
          await runFunctionRunnerWithLogInput(extension, options, JSON.stringify(input), runExport)
        },
        onReloadAndBuildError: async (error) => {
          if (error instanceof FatalError) {
            renderFatalError(error)
          } else {
            outputWarn(`Failed to replay function: ${error.message}`)
          }
        },
        signal: abortController.signal,
      })
    }
  } catch (error) {
    abortController.abort()
    throw error
  }
}

async function runFunctionRunnerWithLogInput(
  fun: ExtensionInstance<FunctionConfigType>,
  options: ReplayOptions,
  input: string,
  exportName: string,
) {
  const outputAsJson = options.json ? ['--json'] : []

  return exec(
    'npm',
    ['exec', '--', 'function-runner', '-f', fun.outputPath, ...outputAsJson, ...['--export', exportName]],
    {
      cwd: fun.directory,
      input,
      stdout: 'inherit',
      stderr: 'inherit',
    },
  )
}

async function getFunctionRunData(functionRunsDir: string, functionHandle: string): Promise<FunctionRunData[]> {
  const allFunctionRunFileNames = readdirSync(functionRunsDir)
    .filter((filename) => {
      // Expected format: 20240522_150641_827Z_extensions_my-function_abcdef.json
      const splitFilename = filename.split('_')
      return splitFilename[3] === 'extensions' && splitFilename[4] === functionHandle
    })
    .reverse()

  let functionRunData: FunctionRunData[] = []
  for (
    let i = 0;
    i < allFunctionRunFileNames.length && functionRunData.length < LOG_SELECTOR_LIMIT;
    i += LOG_SELECTOR_LIMIT
  ) {
    const currentFunctionRunFileNameChunk = allFunctionRunFileNames.slice(i, i + LOG_SELECTOR_LIMIT)
    const functionRunFilePaths = currentFunctionRunFileNameChunk.map((functionRunFile) =>
      joinPath(functionRunsDir, functionRunFile),
    )

    // eslint-disable-next-line no-await-in-loop
    const functionRunDataFromChunk = await Promise.all(
      functionRunFilePaths.map(async (functionRunFilePath) => {
        const fileData = await readFile(functionRunFilePath)
        const parsedData = JSON.parse(fileData)
        return {
          ...parsedData,
          identifier: getIdentifierFromFilename(functionRunFilePath),
        }
      }),
    )

    const filteredFunctionRunDataFromChunk = functionRunDataFromChunk.filter((run) => {
      return run.payload.input != null
    })

    functionRunData = functionRunData.concat(filteredFunctionRunDataFromChunk)
  }

  return functionRunData
}

function getIdentifierFromFilename(fileName: string): string {
  return fileName.split('_').pop()!.substring(0, 6)
}
