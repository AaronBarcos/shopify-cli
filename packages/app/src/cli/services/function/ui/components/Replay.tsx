import {setupExtensionWatcherForReplay} from './hooks/extension-watcher.js'
import {FunctionRunData} from '../../replay.js'
import {ExtensionInstance} from '../../../../models/extensions/extension-instance.js'
import {FunctionConfigType} from '../../../../models/extensions/specifications/function.js'
import {AppInterface} from '../../../../models/app/app.js'
import {prettyPrintJsonIfPossible} from '../../../app-logs/utils.js'
import figures from '@shopify/cli-kit/node/figures'
import {AbortController} from '@shopify/cli-kit/node/abort'
import React, {FunctionComponent} from 'react'
import {Box, Text, Static} from '@shopify/cli-kit/node/ink'

export interface ReplayProps {
  selectedRun: FunctionRunData
  abortController: AbortController
  app: AppInterface
  extension: ExtensionInstance<FunctionConfigType>
}

interface FunctionRun {
  type: 'functionRun'
  input: string
  output: string
  logs: string
  name: string
  size: number
  memory_usage: number
  instructions: number
}

interface SystemMessage {
  type: 'systemMessage'
  message: string
}

type ReplayLog = FunctionRun | SystemMessage

const Replay: FunctionComponent<ReplayProps> = ({selectedRun, abortController, app, extension}) => {
  const {logs, isAborted, canUseShortcuts, statusMessage, recentFunctionRuns, error} = setupExtensionWatcherForReplay({
    selectedRun,
    abortController,
    app,
    extension,
  })

  return (
    <>
      {/* Scrolling upper section */}
      <Static items={logs}>
        {(log, index) => {
          return (
            <Box key={`replayOutputScrollerLog${index}`} flexDirection="column">
              <ReplayLog log={log} />
            </Box>
          )
        }}
      </Static>
      {/* Bottom Bar */}
      {/* eslint-disable-next-line no-negated-condition */}
      {!isAborted ? (
        <Box
          marginY={1}
          paddingTop={1}
          flexDirection="column"
          flexGrow={1}
          borderStyle="single"
          borderBottom={false}
          borderLeft={false}
          borderRight={false}
          borderTop
        >
          {canUseShortcuts ? (
            <Box flexDirection="column">
              <Box flexDirection="row">
                <Text>
                  {figures.pointerSmall} {statusMessage}
                </Text>
              </Box>
              <StatsDisplay recentFunctionRuns={recentFunctionRuns} />
              <Text>
                {figures.pointerSmall} Press <Text bold>q</Text> {figures.lineVertical} quit
              </Text>
            </Box>
          ) : null}
          {error ? <Text color="red">{error}</Text> : null}
        </Box>
      ) : null}
    </>
  )
}

function InputDisplay({input}: {input: string}) {
  return (
    <Box flexDirection="column">
      <Text color="black" backgroundColor="yellow">
        Input
      </Text>
      <Text>{prettyPrintJsonIfPossible(input)}</Text>
    </Box>
  )
}

function LogDisplay({logs}: {logs: string}) {
  return (
    <Box flexDirection="column">
      <Text color="black" backgroundColor="blue">
        Logs
      </Text>
      <Text>{logs}</Text>
    </Box>
  )
}

function OutputDisplay({output}: {output: string}) {
  return (
    <Box flexDirection="column">
      <Text color="black" backgroundColor="green">
        Output
      </Text>
      <Text>{prettyPrintJsonIfPossible(output)}</Text>
    </Box>
  )
}

function BenchmarkDisplay({functionRun}: {functionRun: FunctionRun}) {
  return (
    <Box flexDirection="column">
      <Text color="black" backgroundColor="greenBright">
        Benchmark Results
      </Text>
      <Text>Name: {functionRun.name}</Text>
      <Text>Linear Memory Usage: {functionRun.memory_usage}KB</Text>
      <Text>Instructions: {functionRun.instructions / 1000}K</Text>
      <Text>Size: {functionRun.size}KB</Text>
    </Box>
  )
}

function StatsDisplay({recentFunctionRuns}: {recentFunctionRuns: [FunctionRun, FunctionRun]}) {
  const delta = recentFunctionRuns[0].instructions - recentFunctionRuns[1].instructions
  return (
    <Box flexDirection="column">
      <Text>
        {figures.pointerSmall} Instruction count delta: {delta}
      </Text>
    </Box>
  )
}

function ReplayLog({log}: {log: ReplayLog}) {
  if (log.type === 'functionRun') {
    return (
      <Box flexDirection="column">
        <InputDisplay input={log.input} />
        <LogDisplay logs={log.logs} />
        <OutputDisplay output={log.output} />
        <BenchmarkDisplay functionRun={log} />
      </Box>
    )
  }

  if (log.type === 'systemMessage') {
    return <Text>{log.message}</Text>
  }

  return null
}

export {Replay}
