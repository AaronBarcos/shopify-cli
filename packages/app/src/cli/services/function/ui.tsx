import {render} from '@shopify/cli-kit/node/ui'
import {Replay, ReplayProps} from './ui/components/Replay.js'
import React from 'react'

export async function renderReplay({selectedRun, abortController, app, extension}: ReplayProps) {
  return render(
    <Replay selectedRun={selectedRun} abortController={abortController} app={app} extension={extension} />,
    {
      exitOnCtrlC: false,
    },
  )
}
