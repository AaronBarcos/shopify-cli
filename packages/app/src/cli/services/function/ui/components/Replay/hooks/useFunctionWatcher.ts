import {FunctionRunData} from '../../../../replay.js'
import {AppInterface} from '../../../../../../models/app/app.js'
import {FunctionConfigType} from '../../../../../../models/extensions/specifications/function.js'
import {ExtensionInstance} from '../../../../../../models/extensions/extension-instance.js'
import {setupExtensionWatcher} from '../../../../../dev/extension/bundler.js'
import {FunctionRunFromRunner, ReplayLog} from '../types.js'
import {exec} from '@shopify/cli-kit/node/system'
import {AbortController} from '@shopify/cli-kit/node/abort'
import {useEffect, useRef, useState} from 'react'
import {useInput, useStdin} from '@shopify/cli-kit/node/ink'
import {handleCtrlC} from '@shopify/cli-kit/node/ui'
import {useAbortSignal} from '@shopify/cli-kit/node/ui/hooks'
import {isUnitTest} from '@shopify/cli-kit/node/context/local'
import {treeKill} from '@shopify/cli-kit/node/tree-kill'
import {FatalError} from '@shopify/cli-kit/node/error'
import {Writable} from 'stream'

interface WatchFunctionForReplayOptions {
  selectedRun: FunctionRunData
  abortController: AbortController
  app: AppInterface
  extension: ExtensionInstance<FunctionConfigType>
}

export function useFunctionWatcher({selectedRun, abortController, app, extension}: WatchFunctionForReplayOptions) {
  const functionRunFromSelectedRun = {
    type: 'functionRun',
    input: selectedRun.payload.input,
    output: selectedRun.payload.output,
    logs: selectedRun.payload.logs,
    name: selectedRun.source,
    size: 0,
    memory_usage: 0,
    instructions: selectedRun.payload.fuelConsumed,
  } as FunctionRunFromRunner

  const [logs, setLogs] = useState<ReplayLog[]>([])
  const [recentFunctionRuns, setRecentFunctionRuns] = useState<[FunctionRunFromRunner, FunctionRunFromRunner]>([
    functionRunFromSelectedRun,
    functionRunFromSelectedRun,
  ])

  const [error, setError] = useState<string | undefined>(undefined)

  const {input, export: runExport} = selectedRun.payload

  const {isRawModeSupported: canUseShortcuts} = useStdin()
  const pollingInterval = useRef<NodeJS.Timeout>()

  const [statusMessage, setStatusMessage] = useState(`Watching for changes to ${selectedRun.source}...`)

  useEffect(() => {
    const initialReplay = async () => {
      setStatusMessage('Replaying log with local function...')
      const functionRun = await runFunctionRunnerWithLogInput(extension, JSON.stringify(input), runExport)
      setRecentFunctionRuns((recentFunctionRuns) => {
        return [functionRun, recentFunctionRuns[0]]
      })
      setStatusMessage(`Watching for changes to ${selectedRun.source}...`)
      setLogs((logs) => [...logs, functionRun])
    }

    const startWatchingFunction = async () => {
      const customStdout = new Writable({
        write(chunk, _enconding, next) {
          setLogs((logs) => [...logs, {type: 'systemMessage', message: chunk.toString()}])
          next()
        },
      })

      await setupExtensionWatcher({
        extension,
        app,
        stdout: customStdout,
        stderr: customStdout,
        onChange: async () => {
          setError(undefined)
          setStatusMessage('Re-running with latest changes...')
          const functionRun = await runFunctionRunnerWithLogInput(extension, JSON.stringify(input), runExport)

          setRecentFunctionRuns((recentFunctionRuns) => {
            return [functionRun, recentFunctionRuns[0]]
          })

          setStatusMessage(`Watching for changes to ${selectedRun.source}...`)
          setLogs((logs) => [...logs, functionRun])
        },
        onReloadAndBuildError: async (error) => {
          if (error instanceof FatalError) {
            setError(`Fatal error while reloading and building extension: ${error.formattedMessage || error.message}`)
          } else {
            setError(`Error while reloading and building extension: ${error.message}`)
          }
        },
        signal: abortController.signal,
      })
    }

    // eslint-disable-next-line promise/catch-or-return, @typescript-eslint/no-floating-promises
    initialReplay().then(() => {
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      startWatchingFunction()
    })
  }, [input, runExport, app, extension])

  useAbortSignal(abortController.signal, async (err) => {
    if (err) {
      setStatusMessage('Shutting down replay watcher because of an error ...')
    } else {
      setStatusMessage('Shutting down replay watcher ...')
      setTimeout(() => {
        if (isUnitTest()) return
        treeKill(process.pid, 'SIGINT', false, () => {
          process.exit(0)
        })
      }, 2000)
    }
    clearInterval(pollingInterval.current)
  })

  useInput(
    (input, key) => {
      handleCtrlC(input, key, () => abortController.abort())

      const onInput = async () => {
        try {
          setError('')

          if (input === 'q') {
            abortController.abort()
          }
          // eslint-disable-next-line no-catch-all/no-catch-all
        } catch (_) {
          setError('Failed to handle your input.')
        }
      }

      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      onInput()
    },
    {isActive: Boolean(canUseShortcuts)},
  )
  return {logs, canUseShortcuts, statusMessage, recentFunctionRuns, error}
}

async function runFunctionRunnerWithLogInput(
  fun: ExtensionInstance<FunctionConfigType>,
  input: string,
  exportName: string,
): Promise<FunctionRunFromRunner> {
  let functionRunnerOutput = ''
  const customStdout = new Writable({
    write(chunk, _encoding, next) {
      functionRunnerOutput += chunk
      next()
    },
  })

  await exec('npm', ['exec', '--', 'function-runner', '--json', '-f', fun.outputPath, '--export', exportName], {
    cwd: fun.directory,
    input,
    stdout: customStdout,
    stderr: 'inherit',
  })

  const result = JSON.parse(functionRunnerOutput)
  return {...result, type: 'functionRun'}
}
