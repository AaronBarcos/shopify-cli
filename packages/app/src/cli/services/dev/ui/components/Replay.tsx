import {Box, Text, Static, useInput, useStdin} from '@shopify/cli-kit/node/ink'
import React, {FunctionComponent, useEffect, useMemo, useRef, useState} from 'react'

import figures from '@shopify/cli-kit/node/figures'
import {FunctionRunData} from '../../../function/replay.js'
import {AbortController} from '@shopify/cli-kit/node/abort'
import {setupExtensionWatcher} from '../../extension/bundler.js'
import {exec} from '@shopify/cli-kit/node/system'
import {ExtensionInstance} from '../../../../models/extensions/extension-instance.js'
import {FunctionConfigType} from '../../../../models/extensions/specifications/function.js'
import {AppInterface} from '../../../../models/app/app.js'
import {Writable} from 'stream'
import { prettyPrintJsonIfPossible } from '../../../app-logs/utils.js'
import { handleCtrlC } from '@shopify/cli-kit/node/ui'
import { useAbortSignal } from '@shopify/cli-kit/node/ui/hooks'
import { isUnitTest } from '@shopify/cli-kit/node/context/local'
import { treeKill } from '@shopify/cli-kit/node/tree-kill'

export interface ReplayProps {
  selectedRun: FunctionRunData
  abortController: AbortController
  app: AppInterface
  extension: ExtensionInstance<FunctionConfigType>
}

interface FunctionRun {
  type: 'functionRun'
  input: string
  output: string
  logs: string
  name: string,
  size: number,
  memory_usage: number,
  instructions: number,
}

interface SystemMessage {
  type: 'systemMessage'
  message: string
}

type ReplayLog = FunctionRun | SystemMessage

const Replay: FunctionComponent<ReplayProps> = ({selectedRun, abortController, app, extension}) => {
  const functionRunFromSelectedRun = {
    type: 'functionRun',
    input: selectedRun.payload.input,
    output: selectedRun.payload.output,
    logs: selectedRun.payload.logs,
    name: selectedRun.source,
    size: 0,
    memory_usage: 0,
    instructions: selectedRun.payload.fuelConsumed,
  } as FunctionRun

  // const [functionRuns, setFunctionRuns] = useState<FunctionRun[]>([])
  // const [replayLogs, setReplayLogs] = useState<String[]>([])
  const [logs, setLogs] = useState<ReplayLog[]>([])
  const [recentFunctionRuns, setRecentFunctionRuns] = useState<[FunctionRun, FunctionRun]>([functionRunFromSelectedRun, functionRunFromSelectedRun])

  const [error, setError] = useState<string | undefined>(undefined)

  const {input, export: runExport} = selectedRun.payload


  const {isRawModeSupported: canUseShortcuts} = useStdin()
  const pollingInterval = useRef<NodeJS.Timeout>()

  const delta = useMemo(() => {
    return recentFunctionRuns[0].instructions - recentFunctionRuns[1].instructions
  }, [recentFunctionRuns])

  const [statusMessage, setStatusMessage] = useState("Default Status Message")
  useEffect(() => {
    const startWatchingFunction = async () => {
      const customStdout = new Writable({
        write(chunk, _enconding, next) {
          setLogs((logs) => [...logs, {type: 'systemMessage', message: chunk.toString()}])
          next()
        },
      })

      ;(global as any).andrewStdout = customStdout

      await setupExtensionWatcher({
        extension,
        app,
        stdout: customStdout, // TODO
        stderr: customStdout, // TODO
        onChange: async () => {
          // console.log("in onChange")
          // setLogs((logs) => [...logs, {type: 'systemMessage', message: 'Changes detected, rebuilding and rerunning'}])
          const functionRun = await runFunctionRunnerWithLogInput(extension, JSON.stringify(input), runExport)
          // console.log("++++ the functionRun in onChange")
          // console.log(JSON.parse(functionRun.output).JsonOutput)
          // console.log("the function run to be added")
          // console.log(functionRun)
          // console.log("function after output is swapped")
          // functionRun.output = JSON.parse(functionRun.output).JsonOutput
          // console.log("all the functionRuns")
          // console.log(functionRuns)

          setRecentFunctionRuns((recentFunctionRuns) => {
            return [functionRun, recentFunctionRuns[0]]
          })
          // setPrevLatestFunctionRun(latestFunctionRun)
          // setLatestFunctionRun(functionRun)
          // console.log(">>>> new latest logs")
          // console.log("prev")
          // console.log(prevLatestFunctionRun)
          // console.log("latest")
          // console.log(latestFunctionRun)
          setLogs((logs) => [...logs, functionRun])
        },
        onReloadAndBuildError: async (error) => {
          // TODO: handle error
        },
        signal: abortController.signal,
      })
    }

    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    startWatchingFunction()

    // TODO: return a way to clean up watcher
  }, [input, runExport, app, extension])


  const {isAborted} = useAbortSignal(abortController.signal, async (err) => {
    if (err) {
      setStatusMessage('Shutting down replay watcher because of an error ...')
    } else {
      setStatusMessage('Shutting down replay watcher ...')
      setTimeout(() => {
        if (isUnitTest()) return
        treeKill(process.pid, 'SIGINT', false, () => {
          process.exit(0)
        })
      }, 2000)
    }
    clearInterval(pollingInterval.current)
  })

  useInput(
    (input, key) => {
      handleCtrlC(input, key, () => abortController.abort())

      const onInput = async () => {
        try {
          setError('')

          if (input === 'q') {
            abortController.abort()
          }
          // eslint-disable-next-line no-catch-all/no-catch-all
        } catch (_) {
          setError('Failed to handle your input.')
        }
      }

      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      onInput()
    },
    {isActive: Boolean(canUseShortcuts)},
  )

  return (
    <>
      {/* Scrolling upper section */}
      <Static items={logs}>
        {(log, index) => {
          return (
            <Box key={`randomBoxKey${index}`} flexDirection="column">
              <ReplayLog log={log} />
            </Box>
          )
        }}
      </Static>
      {/* Bottom Bar */}
      <Box
        marginY={1}
        paddingTop={1}
        flexDirection="column"
        flexGrow={1}
        borderStyle="single"
        borderBottom={false}
        borderLeft={false}
        borderRight={false}
        borderTop
      >
        {canUseShortcuts ? (
        <Box flexDirection="column">
          <Box flexDirection="row">
            <Text>
              {figures.pointerSmall}&nbsp;
            </Text>
            {(selectedRun.status === "success") ? (
              <Text color="white" backgroundColor="green">
                {selectedRun.status.toUpperCase()}
              </Text>
              ) : (
              <Text color="white" backgroundColor="red">
                {selectedRun.status.toUpperCase()}
              </Text>
              )}
            <Text>
              &nbsp;| Watching for changes to {selectedRun.source}...
            </Text>
          </Box>
          <Box flexDirection="column">
            <Text>{figures.pointerSmall} Instruction count delta: {delta}</Text>
          </Box>
          <Text>
            {figures.pointerSmall} Press <Text bold>d</Text> {figures.lineVertical} diff output with original
          </Text>
          <Text>
            {figures.pointerSmall} Press <Text bold>q</Text> {figures.lineVertical} quit
          </Text>
        </Box>
        ) : null}
        <Box marginTop={canUseShortcuts ? 1 : 0}>
            <Text>{statusMessage}</Text>
          </Box>
          {error ? <Text color="red">{error}</Text> : null}
      </Box>
    </>
  )
}

function ReplayLog({log}: {log: ReplayLog}) {
  if (log.type === 'functionRun') {
    return (
      <Box flexDirection="column">
        <Text color="black" backgroundColor="yellow">Input</Text>
        <Text>{prettyPrintJsonIfPossible(log.input)}</Text>
        <Text color="black" backgroundColor="blue">Logs</Text>
        <Text>{log.logs}</Text>
        <Text color="black" backgroundColor="green">Output</Text>
        <Text>{prettyPrintJsonIfPossible(log.output)}</Text>
      </Box>
    )
  }

  if (log.type === 'systemMessage') {
    return <Text>{log.message}</Text>
  }

  return null
}

export {Replay}

interface ReplayOptions {
  app: AppInterface
  extension: ExtensionInstance<FunctionConfigType>
  apiKey?: string
  stdout?: boolean
  path: string
  json: boolean
  watch: boolean
  log?: string
}

async function runFunctionRunnerWithLogInput(
  fun: ExtensionInstance<FunctionConfigType>,
  input: string,
  exportName: string,
): Promise<FunctionRun> {
  let functionRunnerOutput = ''
  const customStdout = new Writable({
    write(chunk, _encoding, next) {
      functionRunnerOutput += chunk
      next()
    },
  })

  await exec('npm', ['exec', '--', 'function-runner', '--json', '-f', fun.outputPath, '--export', exportName], {
    cwd: fun.directory,
    input,
    stdout: customStdout,
    stderr: 'inherit',
  })

  const result = JSON.parse(functionRunnerOutput)
  return {...result, type: 'functionRun'}
}
