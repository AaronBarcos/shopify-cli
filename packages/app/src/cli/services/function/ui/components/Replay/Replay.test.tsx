import {Replay} from './Replay.js'
import {useFunctionWatcher} from './hooks/useFunctionWatcher.js'
import {testFunctionExtension, testApp} from '../../../../../models/app/app.test-data.js'
import {ExtensionInstance} from '../../../../../models/extensions/extension-instance.js'
import {FunctionConfigType} from '../../../../../models/extensions/specifications/function.js'
import {FunctionRunData} from '../../../replay.js'
import {AbortController} from '@shopify/cli-kit/node/abort'
import React from 'react'
import {beforeAll, describe, expect, test, vi} from 'vitest'
import {unstyled} from '@shopify/cli-kit/node/output'
import {render} from '@shopify/cli-kit/node/testing/ui'

vi.mock('./hooks/useFunctionWatcher.js')

interface FunctionRun {
  type: 'functionRun'
  input: string
  output: string
  logs: string
  name: string
  size: number
  memory_usage: number
  instructions: number
}

const defaultConfig = {
  name: 'MyFunction',
  type: 'product_discounts',
  build: {
    command: 'make build',
    path: 'dist/index.wasm',
  },
  configuration_ui: true,
  api_version: '2022-07',
  metafields: [],
  handle: 'function-handle',
}

let extension: ExtensionInstance<FunctionConfigType>

beforeAll(async () => {
  extension = await testFunctionExtension({config: defaultConfig})
})

const SELECTED_RUN = {
  shopId: 69665030382,
  apiClientId: 124042444801,
  payload: {
    export: 'run',
    input: {
      someInput: 'someInput',
    },
    inputBytes: 136,
    output: {
      someOutput: 'someOutput',
    },
    outputBytes: 195,
    logs: 'First Log\nLog the second!,\n1,\nfourth line, length should be above!,\nFifth line!',
    functionId: '34236fa9-42f5-4bb6-adaf-956e12fff0b0',
    fuelConsumed: 532632,
  },
  logType: 'function_run',
  cursor: '2024-08-02T17:45:27.683139Z',
  status: 'success',
  source: 'product-discount',
  sourceNamespace: 'extensions',
  logTimestamp: '2024-08-02T17:45:27.382075Z',
  identifier: '123456',
} as FunctionRunData

const FUNCTION_RUN_FROM_SELECTED_RUN = {
  type: 'functionRun',
  input: SELECTED_RUN.payload.input,
  output: SELECTED_RUN.payload.output,
  logs: SELECTED_RUN.payload.logs,
  name: SELECTED_RUN.source,
  size: 0,
  memory_usage: 0,
  instructions: SELECTED_RUN.payload.fuelConsumed,
} as FunctionRun

describe('Replay', () => {
  test('renders a stream of lines from function-runner output, and shortcuts', async () => {
    const watcherReturnValue = {
      logs: [FUNCTION_RUN_FROM_SELECTED_RUN, FUNCTION_RUN_FROM_SELECTED_RUN],
      isAborted: false,
      canUseShortcuts: true,
      statusMessage: `Watching for changes to ${SELECTED_RUN.source}...`,
      recentFunctionRuns: [FUNCTION_RUN_FROM_SELECTED_RUN, FUNCTION_RUN_FROM_SELECTED_RUN],
      error: '',
    }
    const mockedsetupExtensionWatcherForReplay = vi.fn().mockReturnValue(watcherReturnValue)
    vi.mocked(useFunctionWatcher).mockImplementation(mockedsetupExtensionWatcherForReplay)

    const renderInstanceReplay = render(
      <Replay
        selectedRun={SELECTED_RUN}
        abortController={new AbortController()}
        app={testApp()}
        extension={extension}
      />,
    )

    // Then
    expect(unstyled(renderInstanceReplay.lastFrame()!)).toMatchSnapshot()

    // unmount so that polling is cleared after every test
    renderInstanceReplay.unmount()
  })

  test('renders error in the bottom bar when present', async () => {
    const watcherReturnValue = {
      logs: [FUNCTION_RUN_FROM_SELECTED_RUN, FUNCTION_RUN_FROM_SELECTED_RUN],
      isAborted: false,
      canUseShortcuts: true,
      statusMessage: `Watching for changes to ${SELECTED_RUN.source}...`,
      recentFunctionRuns: [FUNCTION_RUN_FROM_SELECTED_RUN, FUNCTION_RUN_FROM_SELECTED_RUN],
      error: 'some error',
    }
    const mockedsetupExtensionWatcherForReplay = vi.fn().mockReturnValue(watcherReturnValue)
    vi.mocked(useFunctionWatcher).mockImplementation(mockedsetupExtensionWatcherForReplay)

    const renderInstanceReplay = render(
      <Replay
        selectedRun={SELECTED_RUN}
        abortController={new AbortController()}
        app={testApp()}
        extension={extension}
      />,
    )

    // Then
    expect(unstyled(renderInstanceReplay.lastFrame()!)).toMatchSnapshot()

    // unmount so that polling is cleared after every test
    renderInstanceReplay.unmount()
  })
})
