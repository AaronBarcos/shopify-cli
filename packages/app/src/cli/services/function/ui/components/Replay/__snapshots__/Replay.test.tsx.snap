// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Replay > renders a stream of lines from function-runner output, and shortcuts 1`] = `
"

            Input            

{
  \\"someInput\\": \\"someInput\\"
}


            Logs            

First Log
Log the second!,
1,
fourth line, length should be above!,
Fifth line!


           Output           

{
  \\"someOutput\\": \\"someOutput\\"
}


      Benchmark Results      

Name: product-discount
Linear Memory Usage: 0KB
Instructions: 532.632K
Size: 0KB


            Input            

{
  \\"someInput\\": \\"someInput\\"
}


            Logs            

First Log
Log the second!,
1,
fourth line, length should be above!,
Fifth line!


           Output           

{
  \\"someOutput\\": \\"someOutput\\"
}


      Benchmark Results      

Name: product-discount
Linear Memory Usage: 0KB
Instructions: 532.632K
Size: 0KB

────────────────────────────────────────────────────────────────────────────────────────────────────

› Watching for changes to product-discount...
› Instruction count delta: 0
› Press q │ quit
"
`;

exports[`Replay > renders error in the bottom bar when present 1`] = `
"

            Input            

{
  \\"someInput\\": \\"someInput\\"
}


            Logs            

First Log
Log the second!,
1,
fourth line, length should be above!,
Fifth line!


           Output           

{
  \\"someOutput\\": \\"someOutput\\"
}


      Benchmark Results      

Name: product-discount
Linear Memory Usage: 0KB
Instructions: 532.632K
Size: 0KB


            Input            

{
  \\"someInput\\": \\"someInput\\"
}


            Logs            

First Log
Log the second!,
1,
fourth line, length should be above!,
Fifth line!


           Output           

{
  \\"someOutput\\": \\"someOutput\\"
}


      Benchmark Results      

Name: product-discount
Linear Memory Usage: 0KB
Instructions: 532.632K
Size: 0KB

────────────────────────────────────────────────────────────────────────────────────────────────────

› Watching for changes to product-discount...
› Instruction count delta: 0
› Press q │ quit
some error
"
`;
